### DISCUSSION TITLE
If `StateProvider` is deprecated in the future, why shouldn't `NotifierProvider` be able to manipulate `state`?

### DISCUSSION THREAD

**Discussion by Eiryyy** [rrousselGit/riverpod#2878] [Q&A]
I recently started using Riverpod. I generated a `NotifierProvider` with `riverpod_generator`, but was surprised that I could not touch the `state`.

```dart
@riverpod
class Email extends _$Email {
  @override
  String build() => '';
}

ref.read(emailProvider.notifier).state = someString;
```

Just to the point, the code above will produce the warning below in `state`.

```
The member 'state' can only be used within 'package:riverpod/src/notifier.dart' or a test.dart[invalid_use_of_visible_for_testing_member](https://dart.dev/diagnostics/invalid_use_of_visible_for_testing_member)
```

I know `StateProvider` will not produce such a warning. But if `StateProvider` is to be deprecated in the future, shouldn't `NotifierProvider` also be able to manipulate state directly?

If I want to cover all providers with code generation, would the workaround be as follows?

```dart
@riverpod
class Email extends _$Email {
  @override
  String build() => '';

  @override
  set state(String value) {
    state = value;
  }
}
```

**Comment by armanafat** [comment]
well you can achive your goal easily and riverpod provides you the tool but you need to do it in right way . 
if you want to access your state of class and change it , as the error suggest you , you need to create function inside of your class and change your state throw that by calling update() method which is available inside your class you can update your state :


@riverpod
class Email extends _$Email {

  @override
  String build() => '';

void changeMyState (String newState){

      update((currentState) => currentState= newState);

}
}

and then in your build method all you need to do is to invoke this function : \

()=>ref.read(emailProvider.notifier).changeMyState(<Your new State which in this case is a String>);

I assume you are familiar with riverpod_generator and build_runner packages and already created <yourClassName>.g.dart file . if not you need to go and read about it .

**Reply by Purus** [reply]
The `update` method does not seem to exist..

**Reply by armanafat** [reply]
according to the document of riverbed itself it was existing but right now idk 
https://riverpod.dev/docs/providers/state_provider

**Comment by gdurandrexel** [comment]
you don't need update(), just `state = newState`.

**Reply by armanafat** [reply]
well ofc in world of programming there is thousands of ways to do something but it's always better to go with the cleanest way there is in order to fallow the clean code principles
